{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Label indices on DataFrames and Series"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When we create a Pandas DataFrame or Series object there is always an associated \"index\" that labels the axes of the DataFrame or Series.  I'll refer to the values of these labels as \"label indices\" to disambiguate them from the standard positional indices that we're used to when working with Python data structures such as lists and arrays.\n",
    "\n",
    "If you don't specify specific label indices, than Pandas will generate a default set of label indices for you:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>YFG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>YFG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  gene\n",
       "0           1  YFG1\n",
       "1           2  YFG2\n",
       "2           3  YFG3\n",
       "3           3  YFG4"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.DataFrame({\n",
    " \"chromosome\": [1, 2, 3, 3],\n",
    " \"gene\": [\"YFG1\", \"YFG2\", \"YFG3\", \"YFG4\"]   \n",
    "})\n",
    "\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RangeIndex(start=0, stop=4, step=1)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.index  # get the label indices associated with the data frame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A Slice generated by taking a single column of a DataFrame shares the same indices:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    YFG1\n",
       "1    YFG2\n",
       "2    YFG3\n",
       "3    YFG4\n",
       "Name: gene, dtype: object"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you want to specify your own indices, use the `index` argument when constructing the data frame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>1</td>\n",
       "      <td>YFG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>2</td>\n",
       "      <td>YFG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  gene\n",
       "a           1  YFG1\n",
       "b           2  YFG2\n",
       "c           3  YFG3\n",
       "d           3  YFG4"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.DataFrame({\n",
    "    \"chromosome\": [1, 2, 3, 3],\n",
    "    \"gene\": [\"YFG1\", \"YFG2\", \"YFG3\", \"YFG4\"]},\n",
    "    index = [\"a\",\"b\",\"c\",\"d\"])\n",
    "\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "a    YFG1\n",
       "b    YFG2\n",
       "c    YFG3\n",
       "d    YFG4\n",
       "Name: gene, dtype: object"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.gene "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Label indices are not necessarily unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>1</td>\n",
       "      <td>YFG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>2</td>\n",
       "      <td>YFG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  gene\n",
       "a           1  YFG1\n",
       "b           2  YFG2\n",
       "c           3  YFG3\n",
       "a           3  YFG4"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1a = pd.DataFrame({\n",
    "    \"chromosome\": [1, 2, 3, 3],\n",
    "    \"gene\": [\"YFG1\", \"YFG2\", \"YFG3\", \"YFG4\"]},\n",
    "    index = [\"a\",\"b\",\"c\",\"a\"])\n",
    "\n",
    "df1a"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Label indices are used when you combine information from different data frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pd.DataFrame(\n",
    "    {\n",
    "    \"chromosome\": [1, 2, 3, 3],\n",
    "    \"gene\": [\"YFG1\", \"YFG2\", \"YFG3\", \"YFG4\"]\n",
    "    },\n",
    ")\n",
    "\n",
    "df2 = pd.DataFrame(\n",
    "    {\n",
    "    \"gene\": [\"YFG4\", \"YFG2\", \"YFG3\", \"YFG1\"],\n",
    "    \"description\": [\"Function X\", \"Function Y\", \"Function Z\", \"\"],\n",
    "    \"reference\": [\"Smith et al. 2010\", \"Magwene 2020\", \"Doe 1984\", \"Botstein 2013\"]\n",
    "    },\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>YFG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>YFG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  gene\n",
       "0           1  YFG1\n",
       "1           2  YFG2\n",
       "2           3  YFG3\n",
       "3           3  YFG4"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>description</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YFG4</td>\n",
       "      <td>Function X</td>\n",
       "      <td>Smith et al. 2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YFG2</td>\n",
       "      <td>Function Y</td>\n",
       "      <td>Magwene 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YFG3</td>\n",
       "      <td>Function Z</td>\n",
       "      <td>Doe 1984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YFG1</td>\n",
       "      <td></td>\n",
       "      <td>Botstein 2013</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene description          reference\n",
       "0  YFG4  Function X  Smith et al. 2010\n",
       "1  YFG2  Function Y       Magwene 2020\n",
       "2  YFG3  Function Z           Doe 1984\n",
       "3  YFG1                  Botstein 2013"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What happens if we try and combine these data frames naively?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>YFG1</td>\n",
       "      <td>Function X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>YFG2</td>\n",
       "      <td>Function Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG3</td>\n",
       "      <td>Function Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG4</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  gene description\n",
       "0           1  YFG1  Function X\n",
       "1           2  YFG2  Function Y\n",
       "2           3  YFG3  Function Z\n",
       "3           3  YFG4            "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1[\"description\"] = df2.description  ## works but gives wrong alignment!\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>description</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>YFG4</td>\n",
       "      <td>Function X</td>\n",
       "      <td>Smith et al. 2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>YFG2</td>\n",
       "      <td>Function Y</td>\n",
       "      <td>Magwene 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>YFG3</td>\n",
       "      <td>Function Z</td>\n",
       "      <td>Doe 1984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>YFG1</td>\n",
       "      <td></td>\n",
       "      <td>Botstein 2013</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene description          reference\n",
       "a  YFG4  Function X  Smith et al. 2010\n",
       "b  YFG2  Function Y       Magwene 2020\n",
       "c  YFG3  Function Z           Doe 1984\n",
       "d  YFG1                  Botstein 2013"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# what happens if data frames have different indices?\n",
    "\n",
    "df2alpha = pd.DataFrame({\n",
    " \"gene\": [\"YFG4\", \"YFG2\", \"YFG3\", \"YFG1\"],\n",
    " \"description\": [\"Function X\", \"Function Y\", \"Function Z\", \"\"],\n",
    " \"reference\": [\"Smith et al. 2010\", \"Magwene 2020\", \"Doe 1984\", \"Botstein 2013\"]\n",
    "},\n",
    "index=[\"a\",\"b\",\"c\",\"d\"])\n",
    "\n",
    "df2alpha"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>YFG1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>YFG2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  gene description\n",
       "0           1  YFG1         NaN\n",
       "1           2  YFG2         NaN\n",
       "2           3  YFG3         NaN\n",
       "3           3  YFG4         NaN"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1[\"description\"] = df2alpha.description  ## works but gives empty values\n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### setting the index after creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pd.DataFrame(\n",
    "    {\n",
    "    \"chromosome\": [1, 2, 3, 3],\n",
    "    \"gene\": [\"YFG1\", \"YFG2\", \"YFG3\", \"YFG4\"]\n",
    "    },\n",
    ")\n",
    "\n",
    "df2 = pd.DataFrame(\n",
    "    {\n",
    "    \"gene\": [\"YFG4\", \"YFG2\", \"YFG3\", \"YFG1\"],\n",
    "    \"description\": [\"Function X\", \"Function Y\", \"Function Z\", \"\"],\n",
    "    \"reference\": [\"Smith et al. 2010\", \"Magwene 2020\", \"Doe 1984\", \"Botstein 2013\"]\n",
    "    },\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YFG1</th>\n",
       "      <td>1</td>\n",
       "      <td>YFG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFG2</th>\n",
       "      <td>2</td>\n",
       "      <td>YFG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFG3</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFG4</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome  gene\n",
       "gene                  \n",
       "YFG1           1  YFG1\n",
       "YFG2           2  YFG2\n",
       "YFG3           3  YFG3\n",
       "YFG4           3  YFG4"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = df1.set_index(df1.gene)\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>description</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YFG4</th>\n",
       "      <td>YFG4</td>\n",
       "      <td>Function X</td>\n",
       "      <td>Smith et al. 2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFG2</th>\n",
       "      <td>YFG2</td>\n",
       "      <td>Function Y</td>\n",
       "      <td>Magwene 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFG3</th>\n",
       "      <td>YFG3</td>\n",
       "      <td>Function Z</td>\n",
       "      <td>Doe 1984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFG1</th>\n",
       "      <td>YFG1</td>\n",
       "      <td></td>\n",
       "      <td>Botstein 2013</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene description          reference\n",
       "gene                                     \n",
       "YFG4  YFG4  Function X  Smith et al. 2010\n",
       "YFG2  YFG2  Function Y       Magwene 2020\n",
       "YFG3  YFG3  Function Z           Doe 1984\n",
       "YFG1  YFG1                  Botstein 2013"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = df2.set_index(df2.gene)\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YFG1</th>\n",
       "      <td>1</td>\n",
       "      <td>YFG1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFG2</th>\n",
       "      <td>2</td>\n",
       "      <td>YFG2</td>\n",
       "      <td>Function Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFG3</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG3</td>\n",
       "      <td>Function Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFG4</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG4</td>\n",
       "      <td>Function X</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome  gene description\n",
       "gene                              \n",
       "YFG1           1  YFG1            \n",
       "YFG2           2  YFG2  Function Y\n",
       "YFG3           3  YFG3  Function Z\n",
       "YFG4           3  YFG4  Function X"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1[\"description\"] = df2[\"description\"]\n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pd.DataFrame(\n",
    "    {\n",
    "    \"chromosome\": [1, 2, 3, 3],\n",
    "    \"gene\": [\"YFG1\", \"YFG2\", \"YFG3\", \"YFG4\"]\n",
    "    },\n",
    ")\n",
    "\n",
    "df2 = pd.DataFrame(\n",
    "    {\n",
    "    \"gene\": [\"YFG4\", \"YFG2\", \"YFG3\", \"YFG1\"],\n",
    "    \"description\": [\"Function X\", \"Function Y\", \"Function Z\", \"\"],\n",
    "    \"reference\": [\"Smith et al. 2010\", \"Magwene 2020\", \"Doe 1984\", \"Botstein 2013\"]\n",
    "    },\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "      <th>description</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>YFG1</td>\n",
       "      <td></td>\n",
       "      <td>Botstein 2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>YFG2</td>\n",
       "      <td>Function Y</td>\n",
       "      <td>Magwene 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG3</td>\n",
       "      <td>Function Z</td>\n",
       "      <td>Doe 1984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG4</td>\n",
       "      <td>Function X</td>\n",
       "      <td>Smith et al. 2010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  gene description          reference\n",
       "0           1  YFG1                  Botstein 2013\n",
       "1           2  YFG2  Function Y       Magwene 2020\n",
       "2           3  YFG3  Function Z           Doe 1984\n",
       "3           3  YFG4  Function X  Smith et al. 2010"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(df1, df2, on=\"gene\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Other types of joins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pd.DataFrame(\n",
    "    {\n",
    "    \"chromosome\": [1, 2, 3, 3, 4],\n",
    "    \"gene\": [\"YFG1\", \"YFG2\", \"YFG3\", \"YFG4\", \"YFG5\"],\n",
    "    },\n",
    ")\n",
    "\n",
    "\n",
    "df2 = pd.DataFrame(\n",
    "    {\n",
    "    \"gene\": [\"YFG4\", \"YFG2\", \"YFG3\", \"YFG1\", \"YFG9\"],\n",
    "    \"description\": [\"Function X\", \"Function Y\", \"Function Z\", \"\", \"Function Q\"],\n",
    "    \"reference\": [\"Smith et al. 2010\", \"Magwene 2020\", \"Doe 1984\", \"Botstein 2013\", \"Dietrich 2004\"]\n",
    "    },\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>YFG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>YFG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>YFG5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  gene\n",
       "0           1  YFG1\n",
       "1           2  YFG2\n",
       "2           3  YFG3\n",
       "3           3  YFG4\n",
       "4           4  YFG5"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>description</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YFG4</td>\n",
       "      <td>Function X</td>\n",
       "      <td>Smith et al. 2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YFG2</td>\n",
       "      <td>Function Y</td>\n",
       "      <td>Magwene 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YFG3</td>\n",
       "      <td>Function Z</td>\n",
       "      <td>Doe 1984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YFG1</td>\n",
       "      <td></td>\n",
       "      <td>Botstein 2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YFG9</td>\n",
       "      <td>Function Q</td>\n",
       "      <td>Dietrich 2004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene description          reference\n",
       "0  YFG4  Function X  Smith et al. 2010\n",
       "1  YFG2  Function Y       Magwene 2020\n",
       "2  YFG3  Function Z           Doe 1984\n",
       "3  YFG1                  Botstein 2013\n",
       "4  YFG9  Function Q      Dietrich 2004"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "      <th>description</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>YFG1</td>\n",
       "      <td></td>\n",
       "      <td>Botstein 2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>YFG2</td>\n",
       "      <td>Function Y</td>\n",
       "      <td>Magwene 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG3</td>\n",
       "      <td>Function Z</td>\n",
       "      <td>Doe 1984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG4</td>\n",
       "      <td>Function X</td>\n",
       "      <td>Smith et al. 2010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  gene description          reference\n",
       "0           1  YFG1                  Botstein 2013\n",
       "1           2  YFG2  Function Y       Magwene 2020\n",
       "2           3  YFG3  Function Z           Doe 1984\n",
       "3           3  YFG4  Function X  Smith et al. 2010"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(df1,df2,on=\"gene\")  # by default merge takes \"inner\" join = intersection of keys from left and right df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "      <th>description</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>YFG1</td>\n",
       "      <td></td>\n",
       "      <td>Botstein 2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>YFG2</td>\n",
       "      <td>Function Y</td>\n",
       "      <td>Magwene 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG3</td>\n",
       "      <td>Function Z</td>\n",
       "      <td>Doe 1984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>YFG4</td>\n",
       "      <td>Function X</td>\n",
       "      <td>Smith et al. 2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>YFG5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  gene description          reference\n",
       "0           1  YFG1                  Botstein 2013\n",
       "1           2  YFG2  Function Y       Magwene 2020\n",
       "2           3  YFG3  Function Z           Doe 1984\n",
       "3           3  YFG4  Function X  Smith et al. 2010\n",
       "4           4  YFG5         NaN                NaN"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(df1, df2, on=\"gene\", how=\"left\") # \"left\" join = keys from left df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "      <th>description</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.0</td>\n",
       "      <td>YFG4</td>\n",
       "      <td>Function X</td>\n",
       "      <td>Smith et al. 2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>YFG2</td>\n",
       "      <td>Function Y</td>\n",
       "      <td>Magwene 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>YFG3</td>\n",
       "      <td>Function Z</td>\n",
       "      <td>Doe 1984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>YFG1</td>\n",
       "      <td></td>\n",
       "      <td>Botstein 2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>YFG9</td>\n",
       "      <td>Function Q</td>\n",
       "      <td>Dietrich 2004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  gene description          reference\n",
       "0         3.0  YFG4  Function X  Smith et al. 2010\n",
       "1         2.0  YFG2  Function Y       Magwene 2020\n",
       "2         3.0  YFG3  Function Z           Doe 1984\n",
       "3         1.0  YFG1                  Botstein 2013\n",
       "4         NaN  YFG9  Function Q      Dietrich 2004"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(df1, df2, on=\"gene\", how=\"right\") # \"right\" join = keys from right df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "      <th>description</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>YFG1</td>\n",
       "      <td></td>\n",
       "      <td>Botstein 2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>YFG2</td>\n",
       "      <td>Function Y</td>\n",
       "      <td>Magwene 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>YFG3</td>\n",
       "      <td>Function Z</td>\n",
       "      <td>Doe 1984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3.0</td>\n",
       "      <td>YFG4</td>\n",
       "      <td>Function X</td>\n",
       "      <td>Smith et al. 2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4.0</td>\n",
       "      <td>YFG5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>YFG9</td>\n",
       "      <td>Function Q</td>\n",
       "      <td>Dietrich 2004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  gene description          reference\n",
       "0         1.0  YFG1                  Botstein 2013\n",
       "1         2.0  YFG2  Function Y       Magwene 2020\n",
       "2         3.0  YFG3  Function Z           Doe 1984\n",
       "3         3.0  YFG4  Function X  Smith et al. 2010\n",
       "4         4.0  YFG5         NaN                NaN\n",
       "5         NaN  YFG9  Function Q      Dietrich 2004"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(df1, df2, on=\"gene\", how=\"outer\")  # \"outer\" join = union of keys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combining information on genome features and gene ontology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "yeast = pd.read_csv(\"/Users/pmagwene/Downloads/SGD_features.tsv\", delimiter=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['SGDID', 'Type', 'Qualifier', 'Name', 'Gene', 'Alias', 'Parent',\n",
       "       'Secondary', 'Chromosome', 'Start', 'Stop', 'Strand', 'GeneticPosition',\n",
       "       'CoordVersion', 'SeqVersion', 'Description'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "yeast.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Yeast GO mapping files\n",
    "\n",
    "go_slim_mapping.tab\tThis file is TAB delimited and contains the mapping of all yeast gene products (protein or RNA)\n",
    "to a GO-Slim term.\n",
    "\n",
    "```\n",
    "1) ORF (mandatory) \t\t- Systematic name of the gene\n",
    "2) Gene (optional) \t\t- Gene name, if one exists\n",
    "3) SGDID (mandatory) \t\t- the SGDID, unique database identifier for the gene\n",
    "4) GO_Aspect (mandatory) \t- which ontology: P=Process, F=Function, C=Component\n",
    "5) GO Slim term (mandatory) \t- the name of the GO term that was selected as a GO Slim term\n",
    "6) GOID (optional) \t\t- the unique numerical identifier of the GO term\n",
    "7) Feature type (mandatory) \t- a description of the sequence feature, such as ORF or tRNA\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "gonames = [\"ORF\", \"Gene\", \"SGDID\", \"GO_aspect\", \"GO_slim\", \"GO_ID\", \"Type\"]\n",
    "goterms = pd.read_table(\"/Users/pmagwene/Downloads/go_slim_mapping.tab\", names=gonames,header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(42885, 7)"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "goterms.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ORF</th>\n",
       "      <th>Gene</th>\n",
       "      <th>SGDID</th>\n",
       "      <th>GO_aspect</th>\n",
       "      <th>GO_slim</th>\n",
       "      <th>GO_ID</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>HRA1</td>\n",
       "      <td>S000119380</td>\n",
       "      <td>C</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0005575</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>HRA1</td>\n",
       "      <td>S000119380</td>\n",
       "      <td>F</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0003674</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>HRA1</td>\n",
       "      <td>S000119380</td>\n",
       "      <td>P</td>\n",
       "      <td>rRNA processing</td>\n",
       "      <td>GO:0006364</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>HRA1</td>\n",
       "      <td>S000119380</td>\n",
       "      <td>P</td>\n",
       "      <td>ribosomal small subunit biogenesis</td>\n",
       "      <td>GO:0042274</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>ICR1</td>\n",
       "      <td>S000132612</td>\n",
       "      <td>C</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>ICR1</td>\n",
       "      <td>S000132612</td>\n",
       "      <td>F</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0003674</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>ICR1</td>\n",
       "      <td>S000132612</td>\n",
       "      <td>P</td>\n",
       "      <td>transcription from RNA polymerase II promoter</td>\n",
       "      <td>GO:0006366</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>LSR1</td>\n",
       "      <td>S000006478</td>\n",
       "      <td>C</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>snRNA_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>LSR1</td>\n",
       "      <td>S000006478</td>\n",
       "      <td>F</td>\n",
       "      <td>RNA binding</td>\n",
       "      <td>GO:0003723</td>\n",
       "      <td>snRNA_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>LSR1</td>\n",
       "      <td>S000006478</td>\n",
       "      <td>P</td>\n",
       "      <td>RNA splicing</td>\n",
       "      <td>GO:0008380</td>\n",
       "      <td>snRNA_gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ORF  Gene       SGDID GO_aspect  \\\n",
       "0  HRA1  HRA1  S000119380         C   \n",
       "1  HRA1  HRA1  S000119380         F   \n",
       "2  HRA1  HRA1  S000119380         P   \n",
       "3  HRA1  HRA1  S000119380         P   \n",
       "4  ICR1  ICR1  S000132612         C   \n",
       "5  ICR1  ICR1  S000132612         F   \n",
       "6  ICR1  ICR1  S000132612         P   \n",
       "7  LSR1  LSR1  S000006478         C   \n",
       "8  LSR1  LSR1  S000006478         F   \n",
       "9  LSR1  LSR1  S000006478         P   \n",
       "\n",
       "                                         GO_slim       GO_ID        Type  \n",
       "0                             cellular_component  GO:0005575  ncRNA_gene  \n",
       "1                             molecular_function  GO:0003674  ncRNA_gene  \n",
       "2                                rRNA processing  GO:0006364  ncRNA_gene  \n",
       "3             ribosomal small subunit biogenesis  GO:0042274  ncRNA_gene  \n",
       "4                                        nucleus  GO:0005634  ncRNA_gene  \n",
       "5                             molecular_function  GO:0003674  ncRNA_gene  \n",
       "6  transcription from RNA polymerase II promoter  GO:0006366  ncRNA_gene  \n",
       "7                                        nucleus  GO:0005634  snRNA_gene  \n",
       "8                                    RNA binding  GO:0003723  snRNA_gene  \n",
       "9                                   RNA splicing  GO:0008380  snRNA_gene  "
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "goterms.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "newfeatures = pd.merge(yeast, goterms, on=\"SGDID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SGDID</th>\n",
       "      <th>Type_x</th>\n",
       "      <th>Qualifier</th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene_x</th>\n",
       "      <th>Alias</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Secondary</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>GeneticPosition</th>\n",
       "      <th>CoordVersion</th>\n",
       "      <th>SeqVersion</th>\n",
       "      <th>Description</th>\n",
       "      <th>ORF</th>\n",
       "      <th>Gene_y</th>\n",
       "      <th>GO_aspect</th>\n",
       "      <th>GO_slim</th>\n",
       "      <th>GO_ID</th>\n",
       "      <th>Type_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S000002143</td>\n",
       "      <td>ORF</td>\n",
       "      <td>Dubious</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chromosome 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>335.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7/31/96</td>\n",
       "      <td>7/31/96</td>\n",
       "      <td>Dubious open reading frame; unlikely to encode...</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0005575</td>\n",
       "      <td>ORF|Dubious</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S000002143</td>\n",
       "      <td>ORF</td>\n",
       "      <td>Dubious</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chromosome 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>335.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7/31/96</td>\n",
       "      <td>7/31/96</td>\n",
       "      <td>Dubious open reading frame; unlikely to encode...</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0003674</td>\n",
       "      <td>ORF|Dubious</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S000002143</td>\n",
       "      <td>ORF</td>\n",
       "      <td>Dubious</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chromosome 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>335.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7/31/96</td>\n",
       "      <td>7/31/96</td>\n",
       "      <td>Dubious open reading frame; unlikely to encode...</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>ORF|Dubious</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows  22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        SGDID Type_x Qualifier     Name Gene_x Alias        Parent Secondary  \\\n",
       "0  S000002143    ORF   Dubious  YAL069W    NaN   NaN  chromosome 1       NaN   \n",
       "1  S000002143    ORF   Dubious  YAL069W    NaN   NaN  chromosome 1       NaN   \n",
       "2  S000002143    ORF   Dubious  YAL069W    NaN   NaN  chromosome 1       NaN   \n",
       "\n",
       "  Chromosome  Start  ...  GeneticPosition CoordVersion  SeqVersion  \\\n",
       "0          1  335.0  ...              NaN      7/31/96     7/31/96   \n",
       "1          1  335.0  ...              NaN      7/31/96     7/31/96   \n",
       "2          1  335.0  ...              NaN      7/31/96     7/31/96   \n",
       "\n",
       "                                         Description      ORF Gene_y  \\\n",
       "0  Dubious open reading frame; unlikely to encode...  YAL069W    NaN   \n",
       "1  Dubious open reading frame; unlikely to encode...  YAL069W    NaN   \n",
       "2  Dubious open reading frame; unlikely to encode...  YAL069W    NaN   \n",
       "\n",
       "  GO_aspect             GO_slim       GO_ID       Type_y  \n",
       "0         C  cellular_component  GO:0005575  ORF|Dubious  \n",
       "1         F  molecular_function  GO:0003674  ORF|Dubious  \n",
       "2         P  biological_process  GO:0008150  ORF|Dubious  \n",
       "\n",
       "[3 rows x 22 columns]"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "newfeatures.head(3)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
